### DISCUSSION TITLE
FIXED: Dynamic configuration values don't work with values that start with a minus (-)

### DISCUSSION THREAD

**Discussion by lm-ydubler** [StackStorm/st2#5392] [General]
It appears there is a bug on the StackStorm side when using [dynamic configuration values](https://docs.stackstorm.com/reference/pack_configs.html#dynamic-configuration-value) -- StackStom can't add a value that starts with a minus sign (-) (and possibly other characters). This is a problem because lots of secret values (such as access keys, etc) can start with a minus sign (-)! It's also possible that other starting characters might result in the same error as well.

Here is the bug submission information:
<img width="661" alt="bug-reporting-info" src="https://user-images.githubusercontent.com/92544319/137533574-3921d204-a21b-46bc-81f1-6d1059aceff7.png">

Here are pictures of the bug occuring:
<img width="922" alt="dynamic-config-values-minus-sign" src="https://user-images.githubusercontent.com/92544319/137533606-0e609b13-c9c5-4bce-ac37-e1ac5a27c36e.png">

Here is the bug occuring when using a live secret access key:
<img width="1076" alt="cant-add-access-key" src="https://user-images.githubusercontent.com/92544319/137533646-0c1661ba-f902-44b3-8f18-f1994b3311a0.png">

Thanks and best.

**Comment by nzlosh** [comment]
This is not a bug but part of how flags are parsed by the CLI.  To supply hyphens at the start of the kvpair value, you must indicate the end of flag processing `--` character sequence. (This is quite common in BASH commands)

### Example 1 - Bare hyphen
```
st2 key set --scope=user kvpair_ok -
+------------------+------------+
| Property         | Value      |
+------------------+------------+
| name             | kvpair_ok |
| value            | -          |
| scope            | st2kv.user |
| expire_timestamp |            |
+------------------+------------+
```
`-` works in this case because it is not a valid flag format so it's treated as a positional argument.

### Example 2 - Hyphen with text starting the kvpair value.
```
st2 key set --scope=user kvpair_fail -ab
usage: st2 [-h] [--version] [--url BASE_URL] [--auth-url AUTH_URL]
           [--api-url API_URL] [--stream-url STREAM_URL]
           [--api-version API_VERSION] [--cacert CACERT]
           [--config-file CONFIG_FILE] [--print-config] [--skip-config]
           [--debug]
           {run,action,action-alias,auth,login,whoami,apikey,execution,inquiry,key,pack,policy,policy-type,rule,webhook,timer,runner,sensor,trace,trigger,trigger-instance,rule-enforcement,workflow,service-registry,role,role-assignment}
           ...
st2: error: unrecognized arguments: -ab
```
In this case, the CLI is processing `-ab` as a flag to be used by `st2` so it raises an error.

### Example 3 - Hyphen with text starting the kvpair value but using the `--` stop flag processing marker.
```
st2 key set --scope=user -- kvpair_ok -ab
+------------------+------------+
| Property         | Value      |
+------------------+------------+
| name             | kvpair_ok |
| value            | -ab        |
| scope            | st2kv.user |
| expire_timestamp |            |
+------------------+------------+
```
In this example, we see `--` stops the cli from processing any text from that point on as a flag for `st2`.  The kv text is taken as positional agruments.

**Reply by lm-ydubler** [reply]
Amazing response, that explains it. Thank you so much, take best care.
